core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -621,7 +621,7 @@ private[log] class Cleaner(val id: Int,
       }
 
       // if we read bytes but didn't get even one complete message, our I/O buffer is too small, grow it and try again
-      if (readBuffer.limit() > 0 && result.messagesRead == 0)
+      if (readBuffer.limit() > 0 && result.messagesRead == 0 && result.batchesDeleted == 0)
```

Comment:
Could we use `bytesRead == 0`?

Reply:
@hachikuji Thanks for the review. For some reason, I thought `bytesRead` may be updated even if no batches were read completely, but looking at the code again, you are right, `bytesRead` should work. Will test and update the PR.